### PR TITLE
Introduce envelope service

### PR DESCRIPTION
**Pull request by doncem**
### JIRA link (if applicable) ###

[Record events in DB](https://tools.hmcts.net/jira/browse/BPS-1040)

### Change description ###

Transactional wrapper for convenient handling of db actions. Some interactions will need to record event as well - naming the test accordingly. At least it won't need to change in future, just update in verification section

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/DataPersister.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        EventRecordRepository eventRecordRepository
+    ) {
+        this.envelopeRepository = envelopeRepository;
+        this.eventRecordRepository = eventRecordRepository;
+    }
+
+    // region BlobProcessor actions
+
+    @Transactional(readOnly = true)
+    public Optional<Envelope> findBlob(String blobName, String containerName) {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/data/DataPersister.java]: `findEnvelope` I guess

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/DataPersister.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    public DataPersister(
+        EnvelopeRepository envelopeRepository,
+        EventRecordRepository eventRecordRepository
+    ) {
+        this.envelopeRepository = envelopeRepository;
+        this.eventRecordRepository = eventRecordRepository;
+    }
+
+    // region BlobProcessor actions
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/data/DataPersister.java]: not sure about those regions, this class shouldn't care who's using its methods imho

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/DataPersister.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import uk.gov.hmcts.reform.blobrouter.data.model.Status;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+import static java.time.Instant.now;
+
+public class DataPersister {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/data/DataPersister.java]: Sounds very generic. Maybe `EnvelopeService`?

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/data/DataPersister.java]: (those event are really envelope-related events)

File: src/test/java/uk/gov/hmcts/reform/blobrouter/data/DataPersisterTest.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+
+@ExtendWith(MockitoExtension.class)
+class DataPersisterTest {
+
+    private static final String BLOB_NAME = "blob";
+    private static final String CONTAINER_NAME = "container";
+    private static final Instant BLOB_CREATED = now();
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/data/DataPersisterTest.java]: I think you can safely ommit `private` here (and for mocks below)

File: src/test/java/uk/gov/hmcts/reform/blobrouter/data/DataPersisterTest.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        // then
+        verify(envelopeRepository).find(BLOB_NAME, CONTAINER_NAME);
+        verifyNoInteractions(eventRecordRepository);
+    }
+
+    @Test
+    void should_create_new_envelope_and_record_event() {
+        // when
+        dataPersister.createDispatchedEnvelope(CONTAINER_NAME, BLOB_NAME, BLOB_CREATED);
+        dataPersister.createRejectedEnvelope(CONTAINER_NAME, BLOB_NAME, BLOB_CREATED);
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/data/DataPersisterTest.java]: worth splitting into 2 separate tests imho

File: src/test/java/uk/gov/hmcts/reform/blobrouter/data/DataPersisterTest.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        verify(eventRecordRepository, never()).insert(newEventRecordCaptor.capture());
+        assertThat(newEventRecordCaptor.getAllValues())
+            .isEmpty();
+    }
+
+    @Test
+    void should_only_call_envelope_repository_to_get_ready_to_delete_blobs() {
+        // when
+        dataPersister.getReadyToDeleteDispatches(CONTAINER_NAME);
+        dataPersister.getReadyToDeleteRejections();
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/data/DataPersisterTest.java]: and here

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/DataPersister.java (hunk 2 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/data/DataPersister.java]: just thought i'd group for user convenience. regions are nor for class/object per se

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/DataPersister.java (hunk 3 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/data/DataPersister.java]: Name is up for discussion - wouldn't go for envelope service though. My random thoughts while generally named as is:

- db command wrapper
- out of the box transaction ability by spring framework + wrapper
- definitely not service
- should live in `data` scope/package

File: src/test/java/uk/gov/hmcts/reform/blobrouter/data/DataPersisterTest.java (hunk 5 above)
Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/blobrouter/data/DataPersisterTest.java]: unnecessary duplication. i think

**Pull request merged**